plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeConnectorTest.java

```diff
@@ -1753,4 +1756,359 @@ private Set<String> getAllFilesFromCdfDirectory(String tableName)
+
+            assertQueryFails(session, "UPDATE " + table.getName() + " SET x = 10 WHERE x = 1", expectedMessageRegExp);
+            assertUpdate(session, "UPDATE " + table.getName() + " SET x = 20 WHERE part = 22", 1);
+
+            assertQueryFails(session, "MERGE INTO " + table.getName() + " t " +
+                                      "USING (SELECT * FROM (VALUES (3, 99), (4,44))) AS s(x, part) " +
+                                      "ON t.x = s.x " +
+                                      "WHEN MATCHED THEN DELETE ", expectedMessageRegExp);
+            assertUpdate(session, "MERGE INTO " + table.getName() + " t " +
+                                  "USING (SELECT * FROM (VALUES (2, 22), (4 , 44))) AS s(x, part) " +
+                                  "ON (t.part = s.part) " +
+                                  "WHEN MATCHED THEN UPDATE " +
+                                  " SET x = t.x + s.x, part = t.part ", 1);
+
+            assertQueryFails(session, "MERGE INTO " + table.getName() + " t " +
+                                      "USING (SELECT * FROM (VALUES (4,44))) AS s(x, part) " +
+                                      "ON t.x = s.x " +
+                                      "WHEN NOT MATCHED THEN INSERT (x, part) VALUES(s.x, s.part) ", expectedMessageRegExp);
+            assertUpdate(session, "MERGE INTO " + table.getName() + " t " +
+                                  "USING (SELECT * FROM (VALUES (4, 44))) AS s(x, part) " +
+                                  "ON (t.part = s.part) " +
+                                  "WHEN NOT MATCHED THEN INSERT (x, part) VALUES(s.x, s.part) ", 1);
+
+            assertQueryFails(session, "DELETE FROM " + table.getName() + " WHERE x = 3", expectedMessageRegExp);
+            assertUpdate(session, "DELETE FROM " + table.getName() + " WHERE part = 33 and x = 3", 1);
+        }
+    }
+
+    @Test
+    public void testPartitionFilterRequiredAndTableChanges()
+    {
+        Session session = sessionWithPartitionFilterRequirement();
+        try (TestTable table = new TestTable(
+                new TrinoSqlExecutor(getQueryRunner(), session),
+                "test_partition_filter_table_changes",
+                "(x INTEGER, part INTEGER) WITH (partitioned_by = ARRAY['part'], change_data_feed_enabled = true)",
+                ImmutableList.of("(1, 11)", "(2, 22)", "(3, 33)"))) {
+            @Language("RegExp")
+            String expectedMessageRegExp = "Filter required on test_schema\\." + table.getName() + " for at least one partition column: part";
+
+            assertQueryFails(session, "UPDATE " + table.getName() + " SET x = 10 WHERE x = 1", expectedMessageRegExp);
+            assertUpdate(session, "UPDATE " + table.getName() + " SET x = 20 WHERE part = 22", 1);
+
+            assertTableChangesQuery("SELECT * FROM TABLE(system.table_changes('test_schema', '" + table.getName() + "'))",
```

Comment:
This query should fail in my opinion.

Reply:
Added testcase for $history table.

Reply:
But `table_changes` table function doesn't support filter pushdown right ?

Reply:
https://github.com/trinodb/trino/pull/17928/files, wip

Reply:
Can we deal with `table_changes` in a follow-up (once #17928 lands)?

Reply:
I think we can after the PR. Let's leave a TODO comment.